### ISSUE TITLE
Tensorflow predict() is accurate on single sample but giving strange results when predicting on dataset

### ISSUE THREAD

**Issue by markstrefford** [tensorflow/tensorflow#56674]
### Issue Type

Support

### Source

binary

### Tensorflow Version

2.8.0

### Custom Code

No

### OS Platform and Distribution

Issue on Windows 11  (with CUDA) and MacOS 12.4 (CPU)

### Python version

3.8 (Windows) and 3.9.5 (MacOS)

### Bazel version

n/a

### GCC/Compiler version

n/a

### CUDA/cuDNN version

CUDA 11.4, driver version 471.41

### GPU model and memory

RTX3070 / 8GB

### Current Behaviour?

```shell
I've trained a model on the Stanford Dogs dataset and get accuracy around 80% (training acc) and 74% (validation acc). 

When I predict using the trained model using `tf_preds = model.predict(test_images)` I get bad predictions (see relevant log output below)

If I iterate through the dataset using a `for` loop, I get very good predictions (see relevant log output below)

I do not understand why this problem occurs. I've raised this on Stackoverflow but had no responses https://stackoverflow.com/questions/72801723/tensorflow-predict-is-accurate-on-single-sample-but-giving-strange-results-whe

I get the same issue when predicting on Windows (env details above) or Mac OS (env details above) with the same trained model.
```

### Standalone code to reproduce the issue

```shell
Here's a link to the notebook in Colab. 

https://colab.research.google.com/drive/1y2-T_D3bQOXbs3tg_UKyYKZVINxb5vF_?usp=sharing

You can visibly see the issue in the confusion matrices at the bottom of the notebook.

Here is the relevant code:


import matplotlib.pyplot as plt
import tensorflow as tf
import tensorflow_datasets as tfds

import numpy as np
from sklearn.metrics import confusion_matrix
import seaborn as sns

dataset, info = tfds.load(name="stanford_dogs", with_info=True)
training_data = dataset['train']
val_data = dataset['test']

IMG_LEN = 224
IMG_SHAPE = (IMG_LEN,IMG_LEN,3)
N_BREEDS = 120

def preprocess(ds_row):
    image = tf.image.convert_image_dtype(ds_row['image'], dtype=tf.float32)
    image = tf.image.resize(image, (IMG_LEN, IMG_LEN), method='nearest')
    
    label = tf.one_hot(ds_row['label'], N_BREEDS)  # TODO: Can remove one_hot and change loss function
    return image, label

def prepare(dataset, batch_size=None):
    ds = dataset.map(preprocess, num_parallel_calls=4)
    ds = ds.shuffle(buffer_size=1000)
    if batch_size:
        ds = ds.batch(batch_size)
        ds = ds.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
    return ds

train_batches = prepare(training_data, batch_size=32)
val_batches = prepare(val_data, batch_size=32)

base_model = tf.keras.applications.MobileNetV2(input_shape = IMG_SHAPE,
                                               include_top=False,
                                               weights='imagenet')

base_model.trainable = False

model = tf.keras.Sequential([
  base_model,
  tf.keras.layers.GlobalAveragePooling2D(),
  tf.keras.layers.Dense(N_BREEDS, activation='softmax')
])

model.compile(optimizer=tf.keras.optimizers.Adamax(0.0001),
              loss='categorical_crossentropy',
              metrics=['accuracy', 'top_k_categorical_accuracy'])

history = model.fit(train_batches,
                    epochs=30,
                    validation_data=val_batches)

# Split out test images and test labels
test_images = dataset['test'].map(
    lambda x:  (tf.image.resize(x['image'], (IMG_LEN, IMG_LEN), method='nearest'))
).batch(1)

_test_labels = dataset['test'].map(
    lambda y: (y['label'])
)
test_labels = [l.numpy() for l in _test_labels]

# Use for loop to get good predictions
_preds = []
_argmax = []
_actuals = []

for dog, label in zip(dataset['test'].take(n), test_labels[:n]):

    pic, _ = preprocess(dog)  # Convert to float32 and resize

    img_tensor = tf.expand_dims(pic,0)
    pred = model(img_tensor)
    
    am = tf.argmax(pred, axis=1)
    _argmax.append(am)

    top_components = tf.reshape(tf.math.top_k(pred, k=5).indices,shape=[-1])
    top_matches = [get_name(i) for i in top_components]
    actual = get_name(label)

    _preds.append(top_components[0])
    _actuals.append(label)

conf_matrix = tf.math.confusion_matrix(
    _actuals,
    _argmax
)

plt.figure(figsize=(20,15))
sns.heatmap(conf_matrix)
plt.ylabel('True label')
plt.xlabel('Predicted label')

# Now using model.predict()
tf_preds = model.predict(test_images)
preds = np.array(tf.math.argmax(tf_preds, axis=1))

conf_matrix = tf.math.confusion_matrix(
    _actuals,
    preds
)

plt.figure(figsize=(20,15))
sns.heatmap(conf_matrix)
plt.ylabel('True label')
plt.xlabel('Predicted label')

```

Note it does not matter whether I run `model.predict()` or the `for` loop first, I get the same result. And I have checked that I am not shuffling the data too (shuffling leads to similar results with `model.predict()`, as in it doesn't help)
```


### Relevant log output

```shell
# using model.predict()
print(test_labels)
[67, 84, 57, 12, 88, 32, 55, 9, 68, 99, 10, 1, 60, 52, 96, 33, 108, 71, 11, 75, 77, 9, 50, 19, 41, 118, 30...]

print(tf_preds)
[12  65   0  65  56  65  91  65  12  50   0  65   0  65   0   0  65  65   58  65   0  65   0   0  65  65  65  65  65  50  65  65  65  97  65  65...]

# using for loop and predicting one image at a time
[67, 84, 57, 12, 88, 32, 55, 9, 68, 99, 10, 1, 60, 52, 96, 33, 108, 71, 11, 75, 77, 9, 50, 19, 41, 118, 30...]

print(_preds)
[ 67  84  57   8  88  32  55   9  68  99  15   1  17  52  96  33 108  71...]
```

**Comment by tilakrayal** [comment]
@sachinprasadhs,
While I was trying to reproduce the provided code, the execution was taking longer than expected. Kindly find the [gist](https://colab.research.google.com/gist/tilakrayal/c6b64b5740e18886f8f9edc52417dc33/2-8dog_classifier.ipynb) of it [here](https://colab.research.google.com/gist/tilakrayal/dcfb4b657c010a678648c3d744498fc5/2-9dog_classifier.ipynb).

**Comment by sachinprasadhs** [comment]
Note that model.predict is not intended to use inside the loop or to iterate data and use it. 
It is intended for batch processing and large number of inputs.
For details check [here](https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#predict).

**Comment by markstrefford** [comment]
> Note that model.predict is not intended to use inside the loop or to iterate data and use it.
> It is intended for batch processing and large number of inputs.
> For details check [here](https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#predict).

Hi yes, I realise that I should use `prediction = model(input_image)` or similar. However, the `for` loop works and predicting on the dataset doesn't
